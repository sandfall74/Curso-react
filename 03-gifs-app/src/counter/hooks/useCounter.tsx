import { useState } from "react";




export const useCounter = (initialValue: number = 10) => {

    const[Counter,setCounter]=useState(initialValue);
const handleAdd=()=>{
    setCounter(Counter+1);  

}

const handleReset=()=>{
    setCounter(initialValue);
}

const handleSubtract=()=>{
    setCounter(pravState=>pravState-1);
}

return {
    //Propiedades
    Counter,
    //metodos
    handleAdd,
    handleReset,
    handleSubtract
}
}
