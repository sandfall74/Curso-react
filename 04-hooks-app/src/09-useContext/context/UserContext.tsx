import {  useEffect, useState, type PropsWithChildren } from "react";
import { users, type User } from "../data/user-mock.data";
import { createContext } from "react";

type AuthStatus = 'checking' | 'authenticated'|'not-autheticated';

interface UserContextProps{
    //state
    authStatus: AuthStatus;
    user: User | null;
    isAuthenticated:boolean;
    //methods
    login:(userId:number)=>boolean;
    logout:()=>void;
}

export const UserContext=createContext({} as UserContextProps);

export const UserContextProvider = ({ children }: PropsWithChildren) => {
const [authStatus,setAuthStatus] = useState<AuthStatus>('checking');
const [user,setUser]=useState<User|null>(null);

const handleLogin = (userId: number) => {
    const user = users.find((user) => user.id === userId);
    if (!user) {
      console.log(`User not found ${userId}`);
      setUser(null);
      setAuthStatus('not-autheticated');
      return false;
    }

    setUser(user);
    setAuthStatus('authenticated');
    localStorage.setItem('userId', userId.toString());
    return true;
  };

const handlelogout =()=>{
    console.log('logout');
    setAuthStatus("not-autheticated");
    setUser(null);
    localStorage.removeItem('userId')
}
useEffect(()=>{
    const storeUserId = localStorage.getItem('userId');
    if(storeUserId){
        handleLogin(+storeUserId);
        return;
    }
    handlelogout();
},[])
  return <UserContext value={{
    authStatus: authStatus,
    user: user,
    isAuthenticated:authStatus==='authenticated',
    login: handleLogin,
    logout: handlelogout
  }}>{children}</UserContext>;
};
