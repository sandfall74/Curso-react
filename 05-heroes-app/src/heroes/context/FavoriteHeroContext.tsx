import {
  createContext,
  useEffect,
  useState,
  type PropsWithChildren,
} from 'react';
import type { Hero } from '../types/hero.interface';

interface FavoriteHeroContext {
  // State
  favorites: Hero[];
  favoriteCount: number;

  // Methods
  isFavorite: (hero: Hero) => boolean;
  toggleFavorite: (hero: Hero) => void;
}

// eslint-disable-next-line react-refresh/only-export-components
export const FavoriteHeroContext = createContext({} as FavoriteHeroContext);

const getFavoritesFromLocalStorage = (): Hero[] => {
  const favorites = localStorage.getItem('favorites');
  return favorites ? JSON.parse(favorites) : [];
};

export const FavoriteHeroProvider = ({ children }: PropsWithChildren) => {
  const [favorites, setFavorites] = useState<Hero[]>(
    getFavoritesFromLocalStorage()
  );

  const toggleFavorite = (hero: Hero) => {
    const heroExist = favorites.find((h) => h.id === hero.id);

    if (heroExist) {
      const newFavorites = favorites.filter((h) => h.id !== hero.id);
      setFavorites(newFavorites);
      return;
    }

    setFavorites([...favorites, hero]);
  };

  const isFavorite = (hero: Hero) => {
    return favorites.some((h) => h.id === hero.id);
  };

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  return (
    <FavoriteHeroContext
      value={{
        // State
        favoriteCount: favorites.length,
        favorites: favorites,
        // Methods
        isFavorite: isFavorite,
        toggleFavorite: toggleFavorite,
      }}
    >
      {children}
    </FavoriteHeroContext>
  );
};