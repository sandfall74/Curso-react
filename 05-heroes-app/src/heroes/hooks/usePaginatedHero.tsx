import { useQuery } from '@tanstack/react-query';
import { getHeroesByPage } from '../actions/get-heroes-by-page.action';

export const usePaginatedHero = (
  page: number,
  limit: number,
  category = 'all'
) => {
  return useQuery({
    queryKey: ['heroes', { page, limit, category }],
    queryFn: () => getHeroesByPage(+page, +limit, category),
    staleTime: 1000 * 60 * 5, // 5 minutos
  });
};
