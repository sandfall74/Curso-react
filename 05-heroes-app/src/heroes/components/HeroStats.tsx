import { use } from 'react';
import { Users, Heart, Zap } from 'lucide-react';
import { HeroStatCard } from './HeroStatCard';

import { Badge } from '@/components/ui/badge';
import { useHeroSummary } from '../hooks/useHeroSummary';
import { FavoriteHeroContext } from '../context/FavoriteHeroContext';

export const HeroStats = () => {
  const { data: summary } = useHeroSummary();
  const { favoriteCount } = use(FavoriteHeroContext);

  if (!summary) {
    return <div>Loading...</div>;
  }

  return (
    <div className="grid grid-cols-2 md:grid-cols-4 gap-4 mb-8">
      <HeroStatCard
        title="Total de personajes"
        icon={<Users className="h-4 w-4 text-muted-foreground" />}
      >
        <div className="text-2xl font-bold">{summary?.totalHeroes}</div>
        <div className="flex gap-1 mt-2">
          <Badge variant="secondary" className="text-xs">
            {summary?.heroCount} Heroes
          </Badge>
          <Badge variant="destructive" className="text-xs">
            {summary?.villainCount} Villains
          </Badge>
        </div>
      </HeroStatCard>

      <HeroStatCard
        title="Favoritos"
        icon={<Heart className="h-4 w-4 text-muted-foreground" />}
      >
        {/* TODO: tenemos que calcular este valor */}
        <div
          className="text-2xl font-bold text-red-600"
          data-testid="favorite-count"
        >
          {favoriteCount}
        </div>
        <p
          className="text-xs text-muted-foreground"
          data-testid="favorite-percentage"
        >
          {((favoriteCount / summary.totalHeroes) * 100).toFixed(2)}% of total
        </p>
      </HeroStatCard>

      <HeroStatCard
        title="Fuerte"
        icon={<Zap className="h-4 w-4 text-muted-foreground" />}
      >
        <div className="text-lg font-bold">{summary?.strongestHero.alias}</div>
        <p className="text-xs text-muted-foreground">
          Strength: {summary?.strongestHero.strength}
        </p>
      </HeroStatCard>

      <HeroStatCard
        title="Inteligente"
        icon={<Heart className="h-4 w-4 text-muted-foreground" />}
      >
        <div className="text-lg font-bold">{summary?.smartestHero.alias}</div>
        <p className="text-xs text-muted-foreground">
          Intelligence: {summary?.smartestHero.intelligence}/10
        </p>
      </HeroStatCard>
    </div>
  );
};