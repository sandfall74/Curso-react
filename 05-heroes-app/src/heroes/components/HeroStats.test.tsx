import { render, screen } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';
import { HeroStats } from './HeroStats';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { useHeroSummary } from '../hooks/useHeroSummary';
import type { SummaryInformationResponse } from '../types/summary-information.response';
import { FavoriteHeroProvider } from '../context/FavoriteHeroContext';

vi.mock('../hooks/useHeroSummary');
const mockUseHeroSummary = vi.mocked(useHeroSummary);

const mockHero = {
  id: '1',
  name: 'Clark Kent',
  slug: 'clark-kent',
  alias: 'Superman',
  powers: [
    'Súper fuerza',
    'Vuelo',
    'Visión de calor',
    'Visión de rayos X',
    'Invulnerabilidad',
    'Súper velocidad',
  ],
  description:
    'El Último Hijo de Krypton, protector de la Tierra y símbolo de esperanza para toda la humanidad.',
  strength: 10,
  intelligence: 8,
  speed: 9,
  durability: 10,
  team: 'Liga de la Justicia',
  image: '1.jpeg',
  firstAppearance: '1938',
  status: 'Active',
  category: 'Hero',
  universe: 'DC',
};

const mockSummaryData: SummaryInformationResponse = {
  totalHeroes: 25,
  strongestHero: {
    id: '1',
    name: 'Clark Kent',
    slug: 'clark-kent',
    alias: 'Superman',
    powers: [
      'Súper fuerza',
      'Vuelo',
      'Visión de calor',
      'Visión de rayos X',
      'Invulnerabilidad',
      'Súper velocidad',
    ],
    description:
      'El Último Hijo de Krypton, protector de la Tierra y símbolo de esperanza para toda la humanidad.',
    strength: 10,
    intelligence: 8,
    speed: 9,
    durability: 10,
    team: 'Liga de la Justicia',
    image: '1.jpeg',
    firstAppearance: '1938',
    status: 'Active',
    category: 'Hero',
    universe: 'DC',
  },
  smartestHero: {
    id: '2',
    name: 'Bruce Wayne',
    slug: 'bruce-wayne',
    alias: 'Batman',
    powers: [
      'Artes marciales',
      'Habilidades de detective',
      'Tecnología avanzada',
      'Sigilo',
      'Genio táctico',
    ],
    description:
      'El Caballero Oscuro de Ciudad Gótica, que utiliza el miedo como arma contra el crimen y la corrupción.',
    strength: 6,
    intelligence: 10,
    speed: 6,
    durability: 7,
    team: 'Liga de la Justicia',
    image: '2.jpeg',
    firstAppearance: '1939',
    status: 'Active',
    category: 'Hero',
    universe: 'DC',
  },
  heroCount: 18,
  villainCount: 7,
};

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const renderHeroStats = (mockData?: Partial<SummaryInformationResponse>) => {
  if (mockData) {
    mockUseHeroSummary.mockReturnValue({
      data: mockData,
    } as unknown as ReturnType<typeof useHeroSummary>);
  } else {
    mockUseHeroSummary.mockReturnValue({
      data: undefined,
    } as unknown as ReturnType<typeof useHeroSummary>);
  }

  return render(
    <QueryClientProvider client={queryClient}>
      <FavoriteHeroProvider>
        <HeroStats />
      </FavoriteHeroProvider>
    </QueryClientProvider>
  );
};

describe('HeroStats', () => {
  test('should render component with default values', () => {
    const { container } = renderHeroStats();

    expect(screen.getByText('Loading...')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should render HeroStats with mock information', () => {
    const { container } = renderHeroStats(mockSummaryData);

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Total de personajes')).toBeDefined();
    expect(screen.getByText('Favoritos')).toBeDefined();
    expect(screen.getByText('Fuerte')).toBeDefined();
  });

  test('should change the percentage of favorites when a hero is added to favorites', () => {
    localStorage.setItem('favorites', JSON.stringify([mockHero]));

    renderHeroStats(mockSummaryData);

    const favoritePercentageElement = screen.getByTestId('favorite-percentage');
    expect(favoritePercentageElement.innerHTML).toContain('4.00%');

    const favoriteCountElement = screen.getByTestId('favorite-count');
    expect(favoriteCountElement.innerHTML).toContain('1');
  });
});
