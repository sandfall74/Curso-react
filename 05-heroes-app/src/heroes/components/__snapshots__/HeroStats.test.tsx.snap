// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeroStats > should render HeroStats with mock information 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 gap-4 mb-8"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Total de personajes
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-users h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <path
            d="M16 3.128a4 4 0 0 1 0 7.744"
          />
          <path
            d="M22 21v-2a4 4 0 0 0-3-3.87"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold"
        >
          25
        </div>
        <div
          class="flex gap-1 mt-2"
        >
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 text-xs"
            data-slot="badge"
          >
            18
             Heroes
          </span>
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 text-xs"
            data-slot="badge"
          >
            7
             Villains
          </span>
        </div>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Favoritos
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-heart h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold text-red-600"
          data-testid="favorite-count"
        >
          0
        </div>
        <p
          class="text-xs text-muted-foreground"
          data-testid="favorite-percentage"
        >
          0.00
          % of total
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Fuerte
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-zap h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 14a1 1 0 0 1-.78-1.63l9.9-10.2a.5.5 0 0 1 .86.46l-1.92 6.02A1 1 0 0 0 13 10h7a1 1 0 0 1 .78 1.63l-9.9 10.2a.5.5 0 0 1-.86-.46l1.92-6.02A1 1 0 0 0 11 14z"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        >
          Superman
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          Strength: 
          10
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Inteligente
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-heart h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        >
          Batman
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          Intelligence: 
          10
          /10
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`HeroStats > should render component with default values 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
